File: apps/web/src/domain/types/maintenancePlan.ts
/**
 * Canonical Domain Interfaces for Maintenance Memberships and Contracts.
 */

export type CanonicalMaintenanceStatus =
  | 'Active'
  | 'Expired'
  | 'Pending Renewal'
  | 'Cancelled'
  | 'On Hold';

export interface CanonicalServiceWindow {
  id: string;
  season?: 'Spring' | 'Summer' | 'Fall' | 'Winter' | 'Bi-Annual' | string;
  windowStartDate?: string;
  windowEndDate?: string;
  dateRange?: string;
  jobSubtitle?: string;
  jobName?: string;
  jobNumber?: string;
  appointmentStatus?: 'Unscheduled' | 'Scheduled' | 'Completed' | string;
  windowNumber?: string; // e.g. "2 of 40"
  lastNotification?: string;
  beginningMonth?: string;
  endMonth?: string;
  isScheduled: boolean;
  isComplete?: boolean;
  appointmentId?: string | null;
  completedDate?: string | null;
  assignedTech?: string | null;
}

export interface CanonicalMaintenancePlan {
  id: string;
  customerId: string;
  customerNumber?: string;
  wexCustomerId?: string;
  customerName?: string;
  name: string; // e.g. "Hvac 1 system maintenance plan"
  status: CanonicalMaintenanceStatus | string;
  description: string;
  startDate: string;
  expiresDate: string;
  expirationDate?: string;
  contractTotal: number;
  annualPrice: number;
  balanceDue: number;
  paymentsApplied?: number;
  contractPrice?: string | number;
  contractBalance?: string | number;
  planPaymentStatus?: string;
  billingFrequency: 'Monthly' | 'Quarterly' | 'Bi-Annual' | 'Annual' | string;
  locationStreet: string;
  locationAddress?: string;
  locationId?: string;
  salesAgent?: string;
  accountName?: string;
  className?: string;
  collectTax?: boolean;
  taxGroup?: string;
  termsText?: string;
  termsAgreed?: boolean;
  customerSignature?: string;
  signatureType?: 'typed' | 'canvas';
  jobType?: string;
  includedVisitsTotal: number;
  includedVisitsRemaining: number;
  coveredEquipmentIds: string[];
  discountPercentage: number; // e.g. 15 for 15% discount
  serviceWindows: CanonicalServiceWindow[];
  sendReminders?: boolean;
  reminderRecipients?: Array<{ name: string; channel: 'Email' | 'Text' | string; destination?: string }>;
  paymentOption?: 'One-Time' | 'Not Now' | string;
  appliedAmount?: number;
  autoRenew: boolean;
  email?: string;
  phone?: string;
  isArchived?: boolean;
  archivedAt?: string | null;
  createdAt: string;
  updatedAt: string;
}

File: apps/mobile/Sources/FSMDemoUI/Models/MaintenancePlan.swift
import Foundation
import Observation
public struct ServiceWindowItem: Identifiable, Hashable, Sendable, Codable {
    public var id = UUID()
    public var season: String
    public var windowStartDate: Date
    public var windowEndDate: Date
    public var isScheduled: Bool
    public var appointmentId: UUID?
    public var dateRange: String
    public var jobSubtitle: String
    public var isComplete: Bool
    public var jobNumber: String?
    public var jobName: String?
    public var appointmentStatus: String?
    public var windowNumber: String?
    public var lastNotification: String?
    
    public init(
        id: UUID = UUID(),
        season: String = "Spring",
        windowStartDate: Date = Date(),
        windowEndDate: Date = Date().addingTimeInterval(86400 * 60),
        isScheduled: Bool = false,
        appointmentId: UUID? = nil,
        dateRange: String = "",
        jobSubtitle: String = "",
        isComplete: Bool = false,
        jobNumber: String? = nil,
        jobName: String? = nil,
        appointmentStatus: String? = nil,
        windowNumber: String? = nil,
        lastNotification: String? = nil
    ) {
        self.id = id
        self.season = season
        self.windowStartDate = windowStartDate
        self.windowEndDate = windowEndDate
        self.isScheduled = isScheduled
        self.appointmentId = appointmentId
        self.dateRange = dateRange.isEmpty ? "\(season) Window" : dateRange
        self.jobSubtitle = jobSubtitle
        self.isComplete = isComplete
        self.jobNumber = jobNumber
        self.jobName = jobName
        self.appointmentStatus = appointmentStatus
        self.windowNumber = windowNumber
        self.lastNotification = lastNotification
    }
    
    public init(dateRange: String, jobSubtitle: String, isComplete: Bool) {
        self.id = UUID()
        self.season = "General"
        self.windowStartDate = Date()
        self.windowEndDate = Date().addingTimeInterval(86400 * 60)
        self.isScheduled = false
        self.appointmentId = nil
        self.dateRange = dateRange
        self.jobSubtitle = jobSubtitle
        self.isComplete = isComplete
        self.jobNumber = nil
        self.jobName = nil
        self.appointmentStatus = nil
        self.windowNumber = nil
        self.lastNotification = nil
    }
}

public struct MaintenancePlanItem: Identifiable, Hashable, Sendable, Codable {
    public var id: UUID
    public var customerId: UUID?
    public var name: String
    public var status: String
    public var description: String
    public var expiresDate: String
    public var contractTotal: String
    public var annualPrice: String
    public var balance: String
    public var locationStreet: String
    public var coveredEquipmentIds: [String]
    public var includedVisitsTotal: Int
    public var includedVisitsRemaining: Int
    public var discountPercentage: Int
    public var serviceWindows: [ServiceWindowItem]
    
    public init(
        id: UUID = UUID(),
        customerId: UUID? = nil,
        name: String = "Gold Protection Plan",
        status: String = "ACTIVE",
        description: String = "Includes 2 bi-annual tune-ups (Spring/Fall) and 15% discount on repair parts.",
        expiresDate: String = "8/12/26",
        contractTotal: String = "$348.00",
        annualPrice: String = "$348.00",
        balance: String = "$174.00",
        locationStreet: String = "",
        coveredEquipmentIds: [String] = [],
        includedVisitsTotal: Int = 2,
        includedVisitsRemaining: Int = 1,
        discountPercentage: Int = 15,
        serviceWindows: [ServiceWindowItem] = []
    ) {
        self.id = id
        self.customerId = customerId
        self.name = name
        self.status = status
        self.description = description
        self.expiresDate = expiresDate
        self.contractTotal = contractTotal
        self.annualPrice = annualPrice
        self.balance = balance
        self.locationStreet = locationStreet
        self.coveredEquipmentIds = coveredEquipmentIds
        self.includedVisitsTotal = includedVisitsTotal
        self.includedVisitsRemaining = includedVisitsRemaining
        self.discountPercentage = discountPercentage
        self.serviceWindows = serviceWindows
    }
    
    public init(
        name: String,
        status: String = "ACTIVE",
        description: String = "Includes 2 bi-annual tune-ups (Spring/Fall) and 15% discount on repair parts.",
        expiresDate: String = "8/12/26",
        contractTotal: String = "$348.00",
        annualPrice: String = "$348.00",
        balance: String = "$174.00",
        locationStreet: String = ""
    ) {
        self.id = UUID()
        self.customerId = nil
        self.name = name
        self.status = status
        self.description = description
        self.expiresDate = expiresDate
        self.contractTotal = contractTotal
        self.annualPrice = annualPrice
        self.balance = balance
        self.locationStreet = locationStreet
        self.coveredEquipmentIds = []
        self.includedVisitsTotal = 2
        self.includedVisitsRemaining = 1
        self.discountPercentage = 15
        self.serviceWindows = []
    }
}
